### DISCUSSION TITLE
How to gradually decrease tempo (rall)?

### DISCUSSION THREAD

**Discussion by Djard** [hydrogen-music/hydrogen#1535] [Hydrogen on Windows]
To me, software manuals focus on the what and not the how, which is useful to know what can be done but not how to do it. Hydrogen's manual is no exception. So I tried to write my own step guide for slowing down tempo in a song. Can someone point out what I am missing in the steps below? I think they are correct, but only to step 6.

 1. Click on "B/C" button (under "Debug" in upper toolbar), until Beat Counter
     appears; or press the comma key.

 2. Click on bar number 7 in the timeline.

 3. Use the +/- buttons at left of the Beat Counter section to set the beat type
     (1/8th, 1/4 etc.). In our example, we have 4/4 time at 232 BPM in 1/16th
     note resolution; so we will set 1/8 notes as the beat type.

 4. Use the +/- buttons at right of the Beat Counter to set 2 as the countdown
     value; i.e. the number of taps before a new tempo is set.  

 5. You may want to tap your foot in time before tapping the comma key to
     start the process. Count as follows, tapping 12 times then pausing when 
     finished. Since our tempo change starts in the middle of bar number 7, we
     will tap at 132 BPM then start to slow down gradually from count number 3.

         COUNT  1  &  2  &  3  &  4  &  5  &  6  &   

 6.  ........ What must we do now to save the programming, since clicking on the 
      Stop play button discards the work?

 7.  Play back the section to audit your work. The original tempo will resume
      from the middle of bar number 8. 

I would rather set tempo markers at points in the BPM ruler, rather than use the tap option; but I do not see how the tempo can be changed from the middle of a measure and end in the middle of another.

**Comment by theGreatWhiteShark** [comment]
Hey @Djard 

> Click on "B/C" button (under "Debug" in upper toolbar), until Beat Counter
appears; or press the comma key.

You can also use Tap Tempo by pressing <kbd>\\</kbd> to adjust the tempo using a moving average of the last hits to the key. But the Beat Counter might indeed be the better choice in here.

<kbd>-</kbd> also decreases tempo but only at a fixed speed (if pressed + hold).

An alternative way to change tempo is to use a MIDI device to send control change signals. This can be e.g. a rotary or fader on a keyboard or a pedal. This one can be hooked to a Hydrogen's MIDI actions in the Preferences, like `BPM_CC_RELATIVE` or `BPM_FINE_CC_RELATIVE`. Moving the controller will then set tempo and give you a finer control.

> What must we do now to save the programming, since clicking on the
Stop play button discards the work?

Sure. Hydrogen does also come with an OSC API. You could write a small script to wait X seconds (time required to reach middle of bar 7) and send OSC messages every X ms to `/Hydrogen/BPM_DECR/`.

> I would rather set tempo markers at points in the BPM ruler, rather than use the tap option; but I do not see how the tempo can be changed from the middle of a measure and end in the middle of another.

Yes. Tempo markers can only be set at the beginning of bars. But you could think of cutting the pattern in half in order to introduce a tempo marker into the middle. Still, this won't result in a gradual decrease of tempo.

**Comment by Djard** [comment]
I don't have a keyboard and prefer not to labor in writing scripts if the manual is correct in asserting that Hydrogen's Tempo Tap feature actually works. Can you not add the missing step below?

 1. Use the +/- buttons at left of the Beat Counter to set the beat type to tap in 1/8 notes.
 2. Use the +/- buttons at right of Beat Counter to set the countdown value as 2, so that 
     a new tempo will be set after every 2 taps.
 3. Select "Song Mode."
 4. Play back the song and start tapping the "\" key at the point where you want the 
     tempo to start slowing down.
 5. Stop tapping at the point where the slowing down of the tempo is to end. Playback
     at the original tempo will resume after this point.
 7. ???

These steps do not work as the tempo remains unchanged; so I must be missing a step. I also have "03" in the Beat Counter display. What does the "03" represent, and how is it changed?

![Beat Counter](https://user-images.githubusercontent.com/99554465/158878822-71d47e3b-ea9b-45e1-8e1e-a5ccd95d6fe3.jpg)

**Reply by theGreatWhiteShark** [reply]
> 6. ???

I'm not quite sure I do understand what is still missing here.

> 5. Stop tapping at the point where the slowing down of the tempo is to end. Playback
at the original tempo will resume after this point.

You can resume with the original tempo by adding a tempo marker at the bar you are done and switch on the Timeline as soon as you are done decreasing the tempo. (Please note that in upcoming versions >=1.2 you can only use beat counter and tap tempo if the Timeline is disabled (BPM button is not highlighted)).

> I also have "03" in the Beat Counter display. What does the "03" represent, and how is it changed?

That's a bug and shouldn't be there. I can reproduce this behavior and can even make Hydrogen crash using the beat counter alone. I'll have a look.

**Comment by cme** [comment]
This feels like the sort of thing that *should* be supported by an Automation Path type control, like the global velocity setting, and would allow a real, gradual rallentando to be added to the song rather than performed manually during playback?

**Reply by theGreatWhiteShark** [reply]
Agreed. And with this in mind I rewrote the audio engine when working on the fix for the tempo markers. From the perspective of the audio engine it _should_ be fairly simple to introduce. Only `computeTickFromFrame()` and `computeFrameFromTick()` have to be made aware of the gradual tempo change.

But the UX of the automation path itself would also need some love.
